neovim: generate init.lua by default

lucc opened the pull request:
Fix #211998. @teto can you please review this and assert that I am really doing what  you propose in #211998.

###### Description of changes

Added a new argument `neovimLuaRcContent` to the wrapper. If neither `neovimLuaRcContent` nor `neovimRcContent` is given the wrapper creates a init.lua file.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

lucc committed:
wrapNeovimUnstable: generate lua rc files by default

Fix #211998.

oddlama commented:
This should also adjust the tests, which still assume init.vim.

https://github.com/NixOS/nixpkgs/blob/9af5c719fa2653e40201bd63e9b158c7669e934b/pkgs/applications/editors/neovim/tests/default.nix#L76

lucc commented:
@oddlama do you know if `neovim.tests.run_nvim_with_ftplugin` is currently broken? For me it hangs on master and on my branch. All other tests seem to pass even without changes to pkgs/applications/editors/neovim/tests/default.nix.

If I change the line you showed from "init.vim" to "init.lua" the situation stays the same.

I assume that my change should not have any effect on existing code as I added a new key to the argument set. The only case where existing code could result in changed behaviour is, when the called did not specify either neovimRcContent or neovimLuaRcContent, in which case an empty config file `init.lua` is generated instead of the empty `init.vim` from before.

The other change is that the given neovim(Lua)RcContent appears in the wrapped `neovim.passthrough.initRc`.

teto commented:
thanks for looking into it. 
One has to keep in mind that sometimes we want to wrap a few settings in the file
https://github.com/NixOS/nixpkgs/blob/c842de8cb731792009dc5efb0e3784b42d5f074e/pkgs/applications/editors/neovim/utils.nix#L111 so the approach I would recommand is to move all to lua, including these wrapped commands and if you detect vimrc settings, then you source a vimscript file as is done in home-manager:
https://github.com/nix-community/home-manager/blob/cf662b6c98a0da81e06066fff0ecf9cbd4627727/modules/programs/neovim.nix#L401

I like that you kept both vimscript and lua (we must for backwards compatibility). The new setting should be documented as well.
If you use home-manager please check this PR doesn't break it (if it does we can update home-manager no problem, but I dont think that's necessary).

lucc commented:
@teto sorry for the wait: I am still at loss about the scope of #211998 and the relation between `nixpkgs#wrapNeovim` and `nixpkgs#wrapNeovimUnstable`. I thought the issue was solely about wrapNeovimUnstable. Can you 
1. explain the interface (or interface changes) you want to either of them?
2. explain the relation between the two?

teto commented:
sry for the confusion, these interfaces need to be documented something but before doing so I wanted to address the issue of lua configuration, (since this month, providers are set via lua, e.g. "vim.g.loaded_${prog}_provider=0";`). Basically wrapNeovim is the legacy one which under the hood relies on wrapNeovimUnstable. As long as I was not sure about wrapNeovimUnstable interface, I didn't want to advertise/create a breaking change. It's a more powerful version of wrapNeovim that is used by home-manager to have more control over the config, but can also be used by nixpkgs users to generate various configs.
I dont have enough cycles to sort this out for nixos-23.05 so I would rather look at this PR after the release.
My main concern is how to handle the transition from the current
 `makeNeovimConfig { plugins = [ { plugin=far-vim; config = "let g:far#source='rg'"; optional = false; } ]; }` to lua. Should we have a hard breaking change and consider that "config" now contains lua code ? this would break everyone's setup. or add `luaConfig` or as in home-manager add "type = 'lua'; (which I've done here https://github.com/NixOS/nixpkgs/pull/170533 I dont like it, it's too complex IMO).

On my list after the release too:
 https://github.com/NixOS/nixpkgs/pull/223172
 
 
 NB: I fixed the tests, they dont hang anymore.

teto commented:
There is a bunch of neovim changes so you may want to wait for those to reach master before doing the following change but I would prefer that we proceed as in home-manager:
- we always write init.lua (depending on wrapRc)
- if there is viml content, we load it from the init.lua, either we source an init.vim in store (as in HM) or you inline it as you prefer.

lucc committed:
wrapNeovimUnstable: source vim init code from a lua init file

This makes the code a little simpler by always adding a lua init file to
the -u option of neovim.

lucc commented:
@teto do you mean #257924 ?

teto commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -30,13 +30,23 @@ let
   let
 
+    selectLua = isNull neovimRcContent;
+    rcContent = if !selectLua then "source ${writeText "init.vim" neovimRcContent}"
```
rcContent should always be neovimLuaRcContent AND `source "init.vim" with neovimRcContent` (order might matter, I would put lua first because I expect it to become the default) this way we dont break existing setups by ignoring the viml content

teto commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -30,13 +30,23 @@ let
     , wrapRc ? true
-    , neovimRcContent ? ""
+    , neovimRcContent ? null
+    , neovimLuaRcContent ? null
```
`neovimLuaRcContent ? ""` makes more sense because we want to call nvim with `-u` even if neovimLuaRcContent is empty because it makes neovim more deterministic by ignoring user config when `wrapRc` is true.

teto replied:
some bikeshedding: the `neovim` prefix in `neovimLuaRcContent` looks redundant since it's a neovim argument already. I know it's consistent with my previous choice neovimRcContent and that I am basically critisizing myself but normally few users would have used it, it's not documented so we could imagine deprecating "neovimRcContent" (I would do it in another PR) so dont let that stop you from choosing another name. at first I had though `initLua` but this is not the full initLua, parts of it is going to be generated + it might source the viml. I guess `luaRcContent` would be fine ? what do you think ?

teto commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -30,13 +30,23 @@ let
+
+  # we can not have both, a lua and a vimL init file, at least one needs to be
+  # null
+  assert isNull neovimRcContent || isNull neovimLuaRcContent;
```
This can be removed since user could set both, if only during the migration towards lua

teto commented:
thinking about these 2 in staging: 
- https://github.com/NixOS/nixpkgs/pull/237425
- https://github.com/NixOS/nixpkgs/pull/260101

lucc committed:
wrapNeovimUnstable: put lua and vimL code into the init file

teto reviewed (approved):
Looks perfect. I'll try it locally then merge I guess.

teto commented:
Would be nice to have an entry saying that neovim now defaults to init.lua in nixos/doc/manual/release-notes/rl-2311.section.md

Also the tests need updating
```
nix-build -A neovim.tests
```
I would recommand changing the configuration to lua, that's probably the most straightforward.

lucc commented:
@teto the failing test uses the old wrapper from`neovimUtils` and that in turn uses `vimUtils`. What do you think is the best course of action?
- add a `customLuaRc` argument to the legacy wrapper
- rewrite the tests to use the new wrapper
- rewrite the assertion in the tests to handle the vim file sourced from the lua init file

lucc committed:
neovim.tests: rename temp file because it now contains lua code

teto commented:
the attrset passed to `makeNeovimConfig` is passed on to the neovim wrapper so if you convert viml's customRC into lua and put it into `luaRcContent`, it should fix most tests. 
We should keep one test that checks the generated viml (sourced by init.lua) script though because most people will keep their viml.

lucc commented:
The tests also use the `plugin.*.config` key to inject viml code from a plugin into the generated init.vim file. Is it in scope for this PR to also add support for a  `plugin.*.luaConfig` key to the arguments of `makeNeovimConfig`?

teto commented:
> The tests also use the plugin.*.config key to inject viml code from a plugin into the generated init.vim file. Is it in scope for this PR to also add support for a plugin.*.luaConfig key to the arguments of makeNeovimConfig?

No. I am not decided with `plugin.*.config` support in nixpkgs, I dont want to commit to more complexity just before the release. We can see if the complexity/benefit is worth the trade off after release.
 
> The problem here is that makeNeovimConfig uses vimUtils.vimrcContent which is obviously not lua-config aware.

yeah I think we can steer away from vimUtils.vimrcContent but after the release since that would be a breaking change (it would mean we dont support plug/pathogen installs anymore).

> This leads me to a more general observation: it seems that all the helper functions and wrappers around neovim use different argument names for similar or the same thing and this seems to be partly because they try to be similar to the vim helper functions. Is it a good idea to try to structure this all a bit better or is it just not worth the work?

I tried to remove some complexity from vim+nixpkgs code (removing support for pathogen/vam etc) while introducing new features and keeping backwards complexity. As a result the nix code to configure neovim is really complex. With structured attributes we have a solution that allows for elegant and simple configurations so I agree with your suggestion that we can improve agument names. But once again we can't do that right before a release so I think what is possible here for 23.11 is to default to init.lua while keeping all the current generation as is (i.e., viml) and just tweak the test to find the viml script referenced by the generated init.lua and compare it against a golden/hardcoded value as is the case.

teto commented:
@lucc now that we have had the branch off for 23.11, we could merge such an update. Would you mind rebasing ? let me know if you have trouble updating the tests.

lucc commented:
I tried to fix the tests some time ago but got stuck and then sidetracked to other things. I will have to dig into it again in order to tell you what I was struggling with.  

If you want you can fix them yourself. Otherwise I will try again some time soon (maybe next week or at the beginning of December).

teto commented:
np but considering all your previous contributions, I am not worried. In any case, let me know here or on matrix, I believe people are excited about this ;)

teto committed:
neovim.tests: fix singe-lines test

teto commented:
I took the liberty of rebasing this PR and fixed the broken test. There is still one broken test https://github.com/NixOS/nixpkgs/issues/209157 but that's either due to neovim or libuv. I have a workaround in my personal config I am hesitant to apply to the tests so it doesn't impact the merge decision.

lucc commented:
Thank you @teto, sorry that I did not come back as soon as promised.

lucc commented:
@linsui it's actually a layer below programs.neovim: programs.neovim.configure uses pkgs.wrapNeovim which internally uses pkgs.wrapNeovimUnstable but the stable wrapper does not expose all options of the unstable wrapper.  But yes it is different layer.

linsui commented:
Yep. The problem is that I still have to write vimrc in vim script and it's loaded from init.lua. This is weird. Maybe we should have a customLuaRC?

linsui commented:
Do you mean you want to remove the module and I have to use an undocumented wrapper? :facepalm:  I thought the module is preferred since it has documents and examples while a wrapper only has source code to read. It's hard to know what I can do and what I should do with a wrapper.
